libp2p/autotls/utils.nim

```diff
@@ -0,0 +1,108 @@
+import
+  base64, strutils, stew/base36, chronos/apps/http/httpclient, chronos, json, net, times
+import
+  ../errors, ../peerid, ../multihash, ../cid, ../multicodec, ../crypto/[crypto, rsa]
+
+type GetPrimaryIPError* = object of LPError
+type AutoTLSError* = object of LPError
+type ACMEError* = object of AutoTLSError
+type PeerIDAuthError* = object of AutoTLSError
+
+const
+  AutoTLSBroker* = "registration.libp2p.direct"
+  AutoTLSDNSServer* = "libp2p.direct"
+  HttpOk* = 200
+  HttpCreated* = 201
+
+proc sampleChar*(ctx: var HmacDrbgContext, choices: string): char =
+  ## Samples a random character from the input string using the DRBG context
+  if choices.len == 0:
+    raise newException(ValueError, "Cannot sample from an empty string")
+  var idx: uint32
+  ctx.generate(idx)
+  return choices[uint32(idx mod uint32(choices.len))]
+
+proc base64UrlEncode*(data: seq[byte]): string =
+  ## Encodes data using base64url (RFC 4648 §5) — no padding, URL-safe
+  var encoded = base64.encode(data, safe = true)
+  encoded.removeSuffix("=")
+  encoded.removeSuffix("=")
+  return encoded
+
+proc isPublicIPv4*(ip: IpAddress): bool =
+  if ip.family != IpAddressFamily.IPv4:
+    return false
+  let ip = $ip
+  return
+    not (
+      ip.startsWith("10.") or
+      (ip.startsWith("172.") and parseInt(ip.split(".")[1]) in 16 .. 31) or
+      ip.startsWith("192.168.") or ip.startsWith("127.") or ip.startsWith("169.254.")
+    )
+
+proc asMoment*(dt: DateTime): Moment =
+  let unixTime: int64 = dt.toTime.toUnix
+  return Moment.init(unixTime, Second)
```

Comment:
This is nice!  Wondering if we should later send a pr to nim-chronos to include this helper